### PR TITLE
fix: QNSPSA with max_evals_grouped exceeding 1

### PR DESCRIPTION
**Pull request by mrossinek**
### Summary

Fixes a bug with QNSPSA when `max_evals_grouped` exceeds 1

### CODE REVIEW

File: qiskit/algorithms/optimizers/qnspsa.py [hunk 1]
```diff
@@ -311,8 +311,18 @@ def get_fidelity(
...
 
+        num_parameters = circuit.num_parameters
+
         def fidelity(values_x, values_y):
-            result = fid.run(circuit, circuit, values_x, values_y).result()
+            values_x = np.reshape(values_x, (-1, num_parameters)).tolist()
+            batch_size_x = len(values_x)
+
+            values_y = np.reshape(values_y, (-1, num_parameters)).tolist()
+            batch_size_y = len(values_y)
```
Comment (jakelishman) [inline qiskit/algorithms/optimizers/qnspsa.py:321]: Do the primitives not accept Numpy array inputs for parameters?  That seems a bit unusual, but this is fine if they don't.

File: test/python/algorithms/optimizers/test_spsa.py [hunk 2]
```diff
@@ -222,3 +222,30 @@ def test_qnspsa_fidelity_primitives(self):
...
+
+        fidelity = QNSPSA.get_fidelity(circuit)
+        optimizer = QNSPSA(fidelity)
+        optimizer.maxiter = 1
+        optimizer.learning_rate = 0.05
+        optimizer.perturbation = 0.05
+        optimizer.set_max_evals_grouped(50)  # greater than 1
+
+        result = optimizer.minimize(objective, initial_point)
+        self.assertIsNotNone(result)
```
Comment (jakelishman) [inline test/python/algorithms/optimizers/test_spsa.py:251]: Could we seed the estimator (just avoiding any possible stochastic convergence failures), and with that, is there anything sensible we can test about the qualities of the output results?  It might be good to test that there are the correct number of them, and that they're within the right ballpark - we can do the quantitative check only if there's a seed, though.

**Review by jakelishman — COMMENTED** [review]
The logic here all looks correct to me, thanks.

File: qiskit/algorithms/optimizers/qnspsa.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/qnspsa.py:321]: That's probably my bad, we used that in the (Sampling)VQE but actually the primitives just need `Sequence[Sequence[float]]` which should work with 2d numpy arrays 🙂

Comment (jakelishman) [reply qiskit/algorithms/optimizers/qnspsa.py:321]: Numpy arrays aren't technically `Sequence` because they don't implement the methods `index` and `count` - it's an unfortunate implication from the `collections.abc` module defining explicit mixin methods as part of the interface, and (as far as I know), there's no built-in Python type that implies the existence of the core iterable/sequence methods, but not the ones with generic definitions.

**Review by jakelishman — APPROVED** [review]
Looks good to me, thanks!
